airbyte-ci/connectors/pipelines/pipelines/actions/environments.py

```diff
@@ -283,6 +283,35 @@ async def find_local_dependencies_in_pyproject_toml(
     return local_dependency_paths
 
 
+def _install_python_dependencies_from_setup_py(
+    container: Container,
+    additional_dependency_groups: Optional[List] = None,
+) -> Container:
+    install_connector_package_cmd = ["python", "-m", "pip", "install", "."]
+    container = container.with_exec(install_connector_package_cmd)
+
+    if additional_dependency_groups:
+        # e.g. .[dev,tests]
+        group_string = f".[{','.join(additional_dependency_groups)}]"
+        group_install_cmd = ["python", "-m", "pip", "install", group_string]
+
+        container = container.with_exec(group_install_cmd)
+
+    return container
+
+
+def _install_python_dependencies_from_requirements_txt(container: Container) -> Container:
+    install_requirements_cmd = ["python", "-m", "pip", "install", "-r", "requirements.txt"]
+    return container.with_exec(install_requirements_cmd)
+
+
+def _install_python_dependencies_from_poetry(container: Container) -> Container:
```

Comment:
Having to install poetry at this level has left me with a thought for when we start removing dockerfiles from python connectors.

The thought is that we have to be really clear in our code and file structure that there is an order of operations.
1. Set base image
2. Install system dependencies
3. Install language libraries

In step 1 we will have to detect
1. Language (java, python, lowcode)
2. Version of language

In step 2 we will have to
1. Provide a method for users to override / add to the system deps in our base image (finalize_build.sh?)

In step 3 we will have to 
1. Detect (or enforce one method) the dependency system (poetry, setup, pipx?, requirements)
2. And install based on that

And we will likely need to have step 3 generic enough to be used to 
1. Build connector images
2. Install one off python utility programs (qa-engine, metadata service, etc)

With all that in mind I think we should start spliting environments.py into a structure based around these different types of interchangable systems like this
```
actions/
    environments/
        utils.py
        system/
            ...
        python/
            pyproject_toml.py
            setup_py.py
            requirements_txt.py
        java/
            ...
```

Anyway no action required. Just wanted to raise the thought

Reply:
@bnchrch we're aligned. 
Just a detail about step 3:
The main change I plan to make is:
* Currently: we create a test environment in a `python` base image on top of which we install stuffs like poetry, pytest etc. I don't like this approach because it means tests are executed in a different environment  from which the actual connector code is executed in the wild (its docker container)
* Target: The build step will continue install all main dependencies. We should install the dev dependencies to run test. So we should just append something like: `pip install .[test]` for `setup.py` or `poetry install --with test`.  This will ensure consistency: tests will be run in the same execution context as the actual "prod" environment.

We should indeed implement a switch like statement to run different command according to the detected package manager in use. (edit: I just saw that you implemented this below)

I'm happy to revisit the `environments` packages (we can also reconsider its legitimacy). I originally named it `environements` to collect reusable containers with pre-installed tools. My target suggestion will discard the provisionning of a `test` environment.